### PR TITLE
Top Level Unions

### PR DESCRIPTION
**Pull request by danstarns**
Adds Unions to the root Query. 

Given the typeDefs: 

```gql
type Movie {
    title: String!
}

type Series {
    name: String!
}

union Production = Movie | Series
```

You can now query: 

```gql
query {
    productions(where?) {
        ... on Movie {
            title
        }
        ... on Series {
            name
        }
    }
}
````

Based on RFC https://github.com/neo4j/graphql/blob/dev/docs/rfcs/rfc-009-top-level-unions.md

## Notes on implementation
Opted to reuse the `translateRead` function and use individual subqueries over the `union` statement as a subquery is more flexible than a union! There is an extra wrapping `CALL` statement and this is so if one union member MATCH is null the entire query won't exit!

### CODE REVIEW

File: packages/graphql/tests/tck/tck-test-files/rfcs/top-level-unions.test.ts [hunk 1]
```diff
@@ -0,0 +1,284 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { gql } from "apollo-server";
+import { Neo4jGraphQL } from "../../../../src";
+import { createJwtRequest } from "../../../utils/create-jwt-request";
+import { formatCypher, translateQuery, formatParams } from "../../utils/tck-test-utils";
+
+describe("integration/rfs/top-level-unions", () => {
```
Comment (tbwiss) [inline packages/graphql/tests/tck/tck-test-files/rfcs/top-level-unions.test.ts:25]: Good old nitpicking...
```suggestion
describe("tck/rfc/top-level-unions", () => {
```

File: packages/graphql/tests/integration/rfcs/top-level-unions.int.test.ts [hunk 2]
```diff
@@ -0,0 +1,362 @@
...
+    let driver: Driver;
+
+    beforeAll(async () => {
+        driver = await neo4j();
+    });
+
+    afterAll(async () => {
+        await driver.close();
+    });
+
```
Comment (tbwiss) [inline packages/graphql/tests/integration/rfcs/top-level-unions.int.test.ts:38]: **question**: Do you think it's worth adding an explicit test for `Two CALL so we can collect and ensure that a NULL MATCH does not break query`? Or is that already tested somewhere and I missed it? in that case ignore this comment.

File: packages/graphql/tests/tck/tck-test-files/rfcs/top-level-unions.test.ts [hunk 3]
```diff
@@ -0,0 +1,284 @@
...
+            }
+            RETURN collect(this_Series2) AS Series
+            }
+            RETURN Movie, Series"
+        `);
+
+        expect(formatParams(result.params)).toMatchInlineSnapshot(`"{}"`);
+    });
+
+    test("should query simple top level unions with where", async () => {
```
Comment (tbwiss) [inline packages/graphql/tests/tck/tck-test-files/rfcs/top-level-unions.test.ts:152]: **question**: In the [RFC](https://github.com/neo4j/graphql/blob/dev/docs/rfcs/rfc-009-top-level-unions.md) it is mentioned that we want to prevent over-fetching when we only return one type. Shall we add a test for that too?

**Review by tbwiss — COMMENTED** [review]
This looks really good!
Left some questions and comments.

**Review by angrykoala — CHANGES_REQUESTED** [review]

File: packages/graphql/src/classes/AbstractGraphElement.ts [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+export interface AbstractGraphElementConstructor {
+    name: string;
+    description?: string;
+}
+
+export abstract class AbstractGraphElement {
```
[the remark this answers is not in the document]
Comment (danstarns) [reply packages/graphql/src/classes/AbstractGraphElement.ts:25]: I designed it to grow and to be used with interfaces too, unions and interfaces are both abstract graph elements.

File: packages/graphql/src/translate/translate-read.ts [hunk 5]
```diff
@@ -45,10 +60,10 @@ function translateRead({ node, context }: { context: Context; node: Node }): [st
     const interfaceStrs: string[] = [];
 
     if (node.queryOptions) {
-        optionsInput.limit = node.queryOptions.getLimit(optionsInput.limit);
+        optionsInput.limit = node.queryOptions.getLimit(optionsInput.limit as number);
```
[the remark this answers is not in the document]
Comment (danstarns) [reply packages/graphql/src/translate/translate-read.ts]: yes we can

File: packages/graphql/src/schema/resolvers/top-level-union-read.ts [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+import getNeo4jResolveTree from "../../utils/get-neo4j-resolve-tree";
+import { Union } from "../../classes/Union";
+
+export default function topLevelUnionReadResolver({ union }: { union: Union }) {
+    async function resolve(_root: any, args: any, _context: unknown, info: GraphQLResolveInfo) {
+        const context = _context as Context;
+
+        context.resolveTree = getNeo4jResolveTree(info, { args });
+
+        const [cypher, params] = translateToplevelUnionRead({ union, context });
```
[the remark this answers is not in the document]
Comment (danstarns) [reply packages/graphql/src/schema/resolvers/top-level-union-read.ts:33]: I was trying to get the same as the relationships but it's very much nested in the projection path, sometimes wrapped in apoc functions and or nested projections. I chose this path because we simply iterate over each member in the union and call `translateRead` thus reusing all of the existing read functionally.
